src/Core/Managed/Shared/TelemetryClientExtensions.cs

```diff
@@ -93,26 +94,78 @@ public static IOperationHolder<T> StartOperation<T>(this TelemetryClient telemet
                 operationTelemetry.GenerateOperationId();
             }
 
+            var telemetryContext = operationTelemetry.Context.Operation;
+
             // If the operation is not executing in the context of any other operation
             // set its name and id as a context (root) operation name and id
-            if (string.IsNullOrEmpty(operationTelemetry.Context.Operation.Id))
+            if (string.IsNullOrEmpty(telemetryContext.Id))
             {
-                operationTelemetry.Context.Operation.Id = operationTelemetry.Id;
+                telemetryContext.Id = operationTelemetry.Id;
             }
 
-            if (string.IsNullOrEmpty(operationTelemetry.Context.Operation.Name))
+            if (string.IsNullOrEmpty(telemetryContext.Name))
             {
-                operationTelemetry.Context.Operation.Name = operationTelemetry.Name;
+                telemetryContext.Name = operationTelemetry.Name;
             }
 
+            bool isActivityEnabled = false;
+#if !NET40
+            if (ActivityExtensions.IsActivityEnabled())
+            {
+                bool operationNameIsSet = false;
+
+                var childActivity = new Activity("Microsoft.ApplicationInsights.OperationContext");
+                if (!string.IsNullOrEmpty(telemetryContext.Name))
+                {
+                    childActivity.SetOperationName(telemetryContext.Name);
+                    operationNameIsSet = true;
+                }
+
+                var parentActivity = Activity.Current;
+
+                if (parentActivity == null)
+                {
+                    if (!string.IsNullOrEmpty(telemetryContext.Id))
+                    {
+                        childActivity.SetParentId(telemetryContext.Id);
+                    }
+                    else if (!string.IsNullOrEmpty(telemetryContext.ParentId))
+                    {
+                        childActivity.SetParentId(telemetryContext.ParentId);
+                    }
+
+                    operationTelemetry.Id = childActivity.ParentId;
```

Comment:
OperationTelemetry.Id is used as a Request-Id (or x-ms-request-id) header.
With Activity/new Http protocol support, it becomes hierarchical.

So, when the incoming request is received, it takes Request-Id and makes it a ParentId of the RequestTelemetry: e.g. 
```
requestTelemetry.Opertion.Context.ParentId = "|guid."
requestTelemetry.Id = "|guid.abc123_"
```

When the dependency call happens, new Activity is created for it (through StartOperation) or directly by Activity.Start(). So the dependency gets an Id:
```
dependencyTelemetry.Opertion.Context.ParentId = "|guid.abc123_"
dependencyTelemetry.Id = "|guid.abc123_1."
```
In case of HTTP, dependencyTelemetry.Id is injected into outgoing request `Request-Id` header.

AppInsights gets root node (`guid`) from each telemetry and sets `Telemetry.Operation.Context.Id` from it.

So what happens here: 
1. Assuming there is no custom Telemetry.Id generation 
if someone has called StartOperation without any parent Activity, operationTelemetry.Id is default one, generated by AppInsgihts in telemetry ctor. We need to align it with the root Operation.Id once we know it (after initialization).
If someone has called StartOperation when there is a parent activity: that activity represents context for this operation, we always make sure Telemetry.Id is correct.

2. Custom Telemetry.Id
I highly doubt there is any reason to override telemetry.Id. @SergeyKanzhelev are you aware of any customer doing this?
 
The idea behind Activity to provide **default**  and **automagical** way to correlate telemetry.
Overriding this behavior breaks the whole correlation feature and we should protect ourselves and customers from it. 
If someone really wants to break default correlation behavior, they can still do it, however I even prefer it to be hard and hacky.